storage/replica.go

```diff
@@ -607,47 +742,70 @@ func (r *Replica) addAdminCmd(ctx context.Context, args proto.Request) (proto.Re
 // addReadOnlyCmd updates the read timestamp cache and waits for any
 // overlapping writes currently processing through Raft ahead of us to
 // clear via the read queue.
-func (r *Replica) addReadOnlyCmd(ctx context.Context, args proto.Request) (proto.Response, error) {
-	header := args.Header()
+func (r *Replica) addReadOnlyCmd(ctx context.Context, bArgs *proto.BatchRequest) (*proto.BatchResponse, error) {
+	header := bArgs.Header()
 
 	if err := r.checkCmdHeader(header); err != nil {
 		return nil, err
 	}
-
-	// If read-consistency is set to INCONSISTENT, run directly.
-	if header.ReadConsistency == proto.INCONSISTENT {
-		// But disallow any inconsistent reads within txns.
-		if header.Txn != nil {
-			return nil, util.Errorf("cannot allow inconsistent reads within a transaction")
-		}
-		if header.Timestamp.Equal(proto.ZeroTimestamp) {
-			header.Timestamp = r.rm.Clock().Now()
-		}
-		reply, intents, err := r.executeCmd(r.rm.Engine(), nil, args)
-		r.handleSkippedIntents(args, intents) // even on error
-		return reply, err
-	} else if header.ReadConsistency == proto.CONSENSUS {
-		return nil, util.Errorf("consensus reads not implemented")
+	if err := r.checkBatchRequest(bArgs); err != nil {
+		return nil, err
 	}
 
+	trace := tracer.FromCtx(ctx)
+
 	// Add the read to the command queue to gate subsequent
 	// overlapping commands until this command completes.
-	cmdKey := r.beginCmd(header, true)
-
-	// This replica must have leader lease to process a consistent read.
-	if err := r.redirectOnOrAcquireLeaderLease(tracer.FromCtx(ctx), header.Timestamp); err != nil {
-		r.endCmd(cmdKey, args, err, true /* readOnly */)
+	qDone := trace.Epoch("command queue")
+	cmdKeys, err := r.beginCmds(bArgs)
+	qDone()
+	if err != nil {
 		return nil, err
 	}
 
-	// Execute read-only command.
-	reply, intents, err := r.executeCmd(r.rm.Engine(), nil, args)
+	// If there are command keys (there might not be if reads are
+	// inconsistent), the read requires the leader lease.
+	if len(cmdKeys) > 0 {
+		if err := r.redirectOnOrAcquireLeaderLease(trace, header.Timestamp); err != nil {
+			r.endCmds(cmdKeys, bArgs, err)
+			return nil, err
+		}
+	}
 
-	// Only update the timestamp cache if the command succeeded.
-	r.endCmd(cmdKey, args, err, true /* readOnly */)
+	// Execute read-only batch command.
+	bReply := &proto.BatchResponse{}
+	bReply.Timestamp = bArgs.Timestamp
+
+	{
```

Comment:
Our treatment of shadowing has gone off the rails somewhere: considered in isolation, shadowing is something to be avoided. We have three principles that are in tension with each other:
1) Avoid shadowing
2) Variable scopes should be no larger than necessary
3) Code should be as straightforward as possible

We have tools that could enforce a complete absence of shadowing, but we have explicitly rejected that level of strictness, because shadowing is often the lesser evil compared to the contortions that are necessary to avoid it. In general, we have opted to use statement-scoped error variables wherever possible. When done consistently (i.e. no error variables are left at the function scope), this avoids shadowing, but we use this pattern *even if it shadows*. Not because shadowing is a good thing, but because we value consistency on this point more than the avoidance of shadowing.

This scope is not idiomatic and introduces shadowing for no good reason. We should either remove this scope or *add* a similar scope around the error declaration at line 760 (I prefer removing this scope). If both errors were in scope blocks then there would be no function-scoped errors, so no shadowing. But if the error from line 760 is still in scope here, we should just use it instead of shadowing it.